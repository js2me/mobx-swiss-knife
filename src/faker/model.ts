import type { Faker } from '@faker-js/faker';
import { computed, makeObservable } from 'mobx';
import type { AnyObject } from 'yummies/types';

import { ModelLoader } from '../model-loader/model-loader.js';

import type { FakerLoaderConfig } from './model.types.js';

declare const process: { env: { NODE_ENV?: string } };

/**
 * [**Documentation**](https://js2me.github.io/mobx-swiss-knife/tools/faker-loader)
 */
export class FakerLoader {
  private modelLoader: ModelLoader<AnyObject>;
  private locale: string;

  constructor(protected config?: FakerLoaderConfig) {
    this.modelLoader = new ModelLoader({
      context: this,
      abortSignal: config?.abortSignal,
    });
    this.locale = this.config?.defaultLocale ?? 'en';

    computed.struct(this, 'isLoading');
    computed.struct(this, 'error');
    makeObservable(this);
  }

  get instance(): Faker {
    const instance = this.modelLoader.get(this.locale);

    if (process.env.NODE_ENV !== 'production' && !instance) {
      throw new Error(
        `Faker instance with locale "${this.locale}" is not found.\n` +
          `Use method load() to load instance`,
      );
    }

    return instance!;
  }

  async load(locale: string = this.locale): Promise<Faker> {
    this.locale = locale;

    await this.modelLoader.load<Faker>(locale, async () => {
      const module = await import(`@faker-js/faker/locale/${locale}`);
      return module.faker;
    });

    return this.modelLoader.get(locale);
  }

  get isLoading() {
    return this.modelLoader.isLoading(this.locale);
  }

  get error() {
    return this.modelLoader.getError(this.locale);
  }

  destroy() {
    this.modelLoader.destroy();
  }
}

/**
 * [**Documentation**](https://js2me.github.io/mobx-swiss-knife/tools/faker-loader)
 */
export const createFakerLoader = (config?: FakerLoaderConfig) =>
  new FakerLoader(config);

/**
 * @deprecated use {FakerLoader}
 */
export const FakerModel = FakerLoader;
