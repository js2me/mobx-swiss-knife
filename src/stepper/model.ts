import {
  action,
  makeObservable,
  observable,
  reaction,
  runInAction,
} from 'mobx';
import { callFunction } from 'yummies/common';

import type { StepperConfig } from './model.types.js';

/**
 * [**Documentation**](https://js2me.github.io/mobx-swiss-knife/tools/stepper)
 */
export class Stepper<StepData> {
  activeStepIndex = 0;

  steps: StepData[] = [];

  get activeStep() {
    return this.steps[this.activeStepIndex];
  }

  constructor({ steps = [], abortSignal }: StepperConfig<StepData>) {
    this.steps = callFunction(steps);

    observable(this, 'activeStepIndex');
    observable(this, 'steps');
    action.bound(this, 'setSteps');
    action.bound(this, 'goToStep');
    action.bound(this, 'nextStep');
    action.bound(this, 'prevStep');
    action.bound(this, 'addStep');
    action.bound(this, 'removeStep');

    makeObservable(this);

    if (typeof steps === 'function') {
      reaction(steps, (steps) => runInAction(() => this.setSteps(steps)), {
        signal: abortSignal,
      });
    }
  }

  setSteps(steps: StepData[]) {
    this.steps = steps;
  }

  goToStep(nextStepIndex: number) {
    this.activeStepIndex = Math.max(
      0,
      Math.min(nextStepIndex, this.steps.length - 1),
    );
  }

  nextStep() {
    this.goToStep(this.activeStepIndex + 1);
  }

  prevStep() {
    this.goToStep(this.activeStepIndex - 1);
  }

  checkStepCompleted(stepIndex: number) {
    return this.activeStepIndex > stepIndex;
  }

  get isNextStepLast() {
    return this.steps.length - 1 === this.activeStepIndex + 1;
  }

  get isLastStep() {
    return this.steps.length - 1 === this.activeStepIndex;
  }

  get hasPrevStep() {
    return this.activeStepIndex !== 0;
  }

  addStep(step: StepData) {
    if (!this.steps.includes(step)) {
      this.steps.push(step);
    }
  }

  removeStep(step: StepData) {
    this.steps = this.steps.filter((it) => it !== step);
  }
}

/**
 * [**Documentation**](https://js2me.github.io/mobx-swiss-knife/tools/stepper)
 */
export const createStepper = <StepData>(config: StepperConfig<StepData>) =>
  new Stepper(config);
