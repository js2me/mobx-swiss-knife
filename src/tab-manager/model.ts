import { action, computed, makeObservable, observable } from 'mobx';

import type { TabManagerConfig, TabManagerItem } from './model.types.js';

/**
 * [**Documentation**](https://js2me.github.io/mobx-swiss-knife/tools/tab-manager)
 */
export class TabManager<T extends TabManagerItem | Readonly<TabManagerItem>> {
  /**
   * This is needed ONLY WHEN `getActiveTab` IS NOT SET
   */
  private localActiveTab!: T['id'];

  private localSettedTabs?: ReadonlyArray<T> | Array<T>;

  get tabs() {
    if (this.localSettedTabs) {
      return this.localSettedTabs;
    }

    if (typeof this.config.tabs === 'function') {
      return this.config.tabs() ?? [];
    }

    return this.config.tabs ?? [];
  }

  protected get tabIndexesMap(): Map<T['id'], number> {
    return new Map(this.tabs.map((tab, i) => [tab.id, i]));
  }

  constructor(private config: TabManagerConfig<T>) {
    observable.ref(this, 'localActiveTab');
    observable.ref(this, 'localSettedTabs');
    action(this, 'setTabs');
    action(this, 'setActiveTab');
    computed.struct(this, 'activeTab');
    computed.struct(this, 'tabs');
    computed(this, 'tabIndexesMap');
    computed.struct(this, 'activeTabData');
    computed.struct(this, 'tabsCount');

    makeObservable(this);
  }

  /**
   * Sets local setted tabs
   * If you are using this method - it means that you are
   * using self controlled tabs.
   * So `tabManager.tabs` will be always equal to `tabs` which
   * you are passed to this method
   *
   * :NOTE: config.tabs function will be ignored!
   */
  setTabs = (tabs: Array<T> | ReadonlyArray<T>) => {
    this.localSettedTabs = tabs;
  };

  getTabData = (tabId: T['id']): T => {
    const index = this.tabIndexesMap.get(tabId)!;
    return this.tabs[index];
  };

  get activeTab(): T['id'] {
    const tabId = this.config.getActiveTab
      ? this.config.getActiveTab()
      : this.localActiveTab;

    const activeTabId = tabId ?? this.config.fallbackTab ?? this.tabs[0].id;

    const tabData = this.getTabData(activeTabId);

    if (!tabData) {
      return this.config.fallbackTab ?? this.tabs[0].id;
    }

    return activeTabId;
  }

  get tabsCount() {
    return this.tabs.length;
  }

  get activeTabData() {
    return this.getTabData(this.activeTab);
  }

  setActiveTab = (activeTabId: T['id']) => {
    if (this.activeTab === activeTabId) {
      return;
    }

    this.config.onChangeActiveTab?.(activeTabId, this.activeTabData);

    if (this.config.getActiveTab == null) {
      this.localActiveTab = activeTabId;
    }
  };

  /**
   * @deprecated nothing to destroy
   */
  destroy() {}
}

/**
 * [**Documentation**](https://js2me.github.io/mobx-swiss-knife/tools/tab-manager)
 */
export const createTabManager = <T extends TabManagerItem>(
  config: TabManagerConfig<T>,
) => new TabManager(config);
