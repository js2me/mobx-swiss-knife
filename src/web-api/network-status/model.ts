import { createAtom, IAtom, reaction } from 'mobx';

import { NetworkStatusParams } from './model.types.js';

export class NetworkStatus {
  private atom: IAtom;
  private disposeFn?: VoidFunction;

  constructor(params?: NetworkStatusParams) {
    this.atom = createAtom(
      process.env.NODE_ENV === 'production' ? '' : 'networkStatusAtom',
      () => globalThis.addEventListener('online', this.atom.reportChanged),
      () => globalThis.removeEventListener('online', this.atom.reportChanged),
    );

    if (params?.whenOnline || params?.whenOffline) {
      this.disposeFn = reaction(
        () => this.isOnline,
        (isOnline) => {
          if (isOnline) {
            params?.whenOnline?.();
          } else {
            params?.whenOffline?.();
          }
        },
        { signal: params?.abortSignal, fireImmediately: true },
      );
    }
  }

  get isOnline() {
    this.atom.reportObserved();
    return navigator.onLine;
  }

  get isOffline() {
    return !this.isOnline;
  }

  protected handleOnlineChange = () => {
    this.atom.reportChanged();
  };

  destroy() {
    this.atom.observers_.clear();
    this.disposeFn?.();
  }
}

export const createNetworkStatus = (params?: NetworkStatusParams) =>
  new NetworkStatus(params);
