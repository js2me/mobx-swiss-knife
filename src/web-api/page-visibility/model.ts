/* eslint-disable unicorn/consistent-function-scoping */
import { createAtom, IAtom, reaction } from 'mobx';

import { PageVisibilityParams } from './model.types.js';

export class PageVisibility {
  private disposeFn?: VoidFunction;

  private static atom?: IAtom;

  constructor(params?: PageVisibilityParams) {
    if (params?.whenVisible || params?.whenHidden) {
      this.disposeFn = reaction(
        () => this.isVisible,
        (isVisible) => {
          if (isVisible) {
            params?.whenVisible?.();
          } else {
            params?.whenHidden?.();
          }
        },
        { signal: params?.abortSignal, fireImmediately: true },
      );
    }
  }

  get isVisible() {
    return PageVisibility.isVisible;
  }

  get isHidden() {
    return PageVisibility.isHidden;
  }

  static get isVisible() {
    if (PageVisibility.atom === undefined) {
      PageVisibility.atom = createAtom(
        process.env.NODE_ENV === 'production' ? '' : 'pageVisibilityAtom',
        () =>
          globalThis.addEventListener(
            'visibilitychange',
            PageVisibility.atom!.reportChanged,
          ),
        () =>
          globalThis.removeEventListener(
            'visibilitychange',
            PageVisibility.atom!.reportChanged,
          ),
      );
    }
    PageVisibility.atom.reportObserved();
    return document.visibilityState === 'visible';
  }

  static get isHidden() {
    return !this.isVisible;
  }

  destroy() {
    this.disposeFn?.();
  }
}

export const createPageVisibility = (params?: PageVisibilityParams) =>
  new PageVisibility(params);
