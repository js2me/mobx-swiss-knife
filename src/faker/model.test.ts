import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';

import { FakerLoader, createFakerLoader } from './model.js';
import { FakerLoaderConfig } from './model.types.js';

describe('FakerLoader', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  describe('constructor', () => {
    it('should initialize with default values', () => {
      const loader = new FakerLoader();

      expect(loader).toBeInstanceOf(FakerLoader);
    });

    it('should initialize with config', () => {
      const config: FakerLoaderConfig = { defaultLocale: 'fr' };
      const loader = new FakerLoader(config);

      expect(loader).toBeInstanceOf(FakerLoader);
    });

    it('should create a model loader instance', () => {
      const loader = new FakerLoader();

      expect(loader).toHaveProperty('modelLoader');
    });
  });

  describe('createFakerLoader factory function', () => {
    it('should create a new FakerLoader instance', () => {
      const config: FakerLoaderConfig = {};
      const loader = createFakerLoader(config);

      expect(loader).toBeInstanceOf(FakerLoader);
    });
  });
});
