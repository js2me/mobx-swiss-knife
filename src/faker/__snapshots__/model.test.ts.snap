// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FakerLoader > instance getter > should throw an error in development mode when instance is not found 1`] = `
[Error: Faker instance with locale "en" is not found.
Use method load() to load instance]
`;
